

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>HiVis documentation &mdash; HiVis 0.1.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=01f34227"></script>
      <script src="_static/doctools.js?v=9bcbadda"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Aggregation" href="test.html" />
    <link rel="prev" title="HD Integrated Visium Interactive Suite (HiVis)" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            HiVis
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">HiVis documentation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#module-HiVis">HiVis</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#HiVis.load"><code class="docutils literal notranslate"><span class="pre">load()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#HiVis.new"><code class="docutils literal notranslate"><span class="pre">new()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#HiVis.HiVis"><code class="docutils literal notranslate"><span class="pre">HiVis</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#HiVis.HiVis.recolor"><code class="docutils literal notranslate"><span class="pre">HiVis.recolor()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#HiVis.HiVis.qc"><code class="docutils literal notranslate"><span class="pre">HiVis.qc()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#HiVis.HiVis.add_mask"><code class="docutils literal notranslate"><span class="pre">HiVis.add_mask()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#HiVis.HiVis.add_annotations"><code class="docutils literal notranslate"><span class="pre">HiVis.add_annotations()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#HiVis.HiVis.dge"><code class="docutils literal notranslate"><span class="pre">HiVis.dge()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#HiVis.HiVis.add_agg"><code class="docutils literal notranslate"><span class="pre">HiVis.add_agg()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#HiVis.HiVis.agg_stardist"><code class="docutils literal notranslate"><span class="pre">HiVis.agg_stardist()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#HiVis.HiVis.add_meta"><code class="docutils literal notranslate"><span class="pre">HiVis.add_meta()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#HiVis.HiVis.update_meta"><code class="docutils literal notranslate"><span class="pre">HiVis.update_meta()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#HiVis.HiVis.pseudobulk"><code class="docutils literal notranslate"><span class="pre">HiVis.pseudobulk()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#HiVis.HiVis.noise_mean_curve"><code class="docutils literal notranslate"><span class="pre">HiVis.noise_mean_curve()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#HiVis.HiVis.cor"><code class="docutils literal notranslate"><span class="pre">HiVis.cor()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#HiVis.HiVis.export_h5"><code class="docutils literal notranslate"><span class="pre">HiVis.export_h5()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#HiVis.HiVis.export_images"><code class="docutils literal notranslate"><span class="pre">HiVis.export_images()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#HiVis.HiVis.get"><code class="docutils literal notranslate"><span class="pre">HiVis.get()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#HiVis.HiVis.subset"><code class="docutils literal notranslate"><span class="pre">HiVis.subset()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#HiVis.HiVis.remove_pixels"><code class="docutils literal notranslate"><span class="pre">HiVis.remove_pixels()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#HiVis.HiVis.head"><code class="docutils literal notranslate"><span class="pre">HiVis.head()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#HiVis.HiVis.shape"><code class="docutils literal notranslate"><span class="pre">HiVis.shape</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#HiVis.HiVis.columns"><code class="docutils literal notranslate"><span class="pre">HiVis.columns</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#HiVis.HiVis.rename"><code class="docutils literal notranslate"><span class="pre">HiVis.rename()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#HiVis.HiVis.update"><code class="docutils literal notranslate"><span class="pre">HiVis.update()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#HiVis.HiVis.copy"><code class="docutils literal notranslate"><span class="pre">HiVis.copy()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#HiVis.HiVis.save"><code class="docutils literal notranslate"><span class="pre">HiVis.save()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-Aggregation">Aggregation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Aggregation.Aggregation"><code class="docutils literal notranslate"><span class="pre">Aggregation</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Aggregation.Aggregation.import_geometry"><code class="docutils literal notranslate"><span class="pre">Aggregation.import_geometry()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#Aggregation.Aggregation.merge"><code class="docutils literal notranslate"><span class="pre">Aggregation.merge()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#Aggregation.Aggregation.get"><code class="docutils literal notranslate"><span class="pre">Aggregation.get()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#Aggregation.Aggregation.subset"><code class="docutils literal notranslate"><span class="pre">Aggregation.subset()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#Aggregation.Aggregation.pseudobulk"><code class="docutils literal notranslate"><span class="pre">Aggregation.pseudobulk()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#Aggregation.Aggregation.smooth"><code class="docutils literal notranslate"><span class="pre">Aggregation.smooth()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#Aggregation.Aggregation.noise_mean_curve"><code class="docutils literal notranslate"><span class="pre">Aggregation.noise_mean_curve()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#Aggregation.Aggregation.cor"><code class="docutils literal notranslate"><span class="pre">Aggregation.cor()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#Aggregation.Aggregation.sync"><code class="docutils literal notranslate"><span class="pre">Aggregation.sync()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#Aggregation.Aggregation.export_h5"><code class="docutils literal notranslate"><span class="pre">Aggregation.export_h5()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#Aggregation.Aggregation.dge"><code class="docutils literal notranslate"><span class="pre">Aggregation.dge()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#Aggregation.Aggregation.shape"><code class="docutils literal notranslate"><span class="pre">Aggregation.shape</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#Aggregation.Aggregation.combine"><code class="docutils literal notranslate"><span class="pre">Aggregation.combine()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#Aggregation.Aggregation.update"><code class="docutils literal notranslate"><span class="pre">Aggregation.update()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#Aggregation.Aggregation.head"><code class="docutils literal notranslate"><span class="pre">Aggregation.head()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#Aggregation.Aggregation.columns"><code class="docutils literal notranslate"><span class="pre">Aggregation.columns</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#Aggregation.Aggregation.copy"><code class="docutils literal notranslate"><span class="pre">Aggregation.copy()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#Aggregation.Aggregation.rename"><code class="docutils literal notranslate"><span class="pre">Aggregation.rename()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#Aggregation.Aggregation.export_to_matlab"><code class="docutils literal notranslate"><span class="pre">Aggregation.export_to_matlab()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="test.html">Aggregation</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">HiVis</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">HiVis documentation</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/modules.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="hivis-documentation">
<h1>HiVis documentation<a class="headerlink" href="#hivis-documentation" title="Link to this heading"></a></h1>
<section id="module-HiVis">
<span id="hivis"></span><h2>HiVis<a class="headerlink" href="#module-HiVis" title="Link to this heading"></a></h2>
<p>HD Integrated Visium Interactive Suite (HiVis)</p>
<dl class="py function">
<dt class="sig sig-object py" id="HiVis.load">
<span class="sig-prename descclassname"><span class="pre">HiVis.</span></span><span class="sig-name descname"><span class="pre">load</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directory</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/HiVis.html#load"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#HiVis.load" title="Link to this definition"></a></dt>
<dd><p>loads an instance from a pickle format
parameters:</p>
<blockquote>
<div><ul class="simple">
<li><p>filename - full path of pkl file, or just the filename if directory is specified</p></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="HiVis.new">
<span class="sig-prename descclassname"><span class="pre">HiVis.</span></span><span class="sig-name descname"><span class="pre">new</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path_image_fullres</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_input_data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_output</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crop_images</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">properties</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">on_tissue_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_reads_in_spot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_reads_gene</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fluorescence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_qc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/HiVis.html#new"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#HiVis.new" title="Link to this definition"></a></dt>
<dd><ul class="simple">
<li><p>Loads images, data and metada.</p></li>
<li><p>Initializes the connection from the data and metadata to the images coordinates</p></li>
<li><p>Adds basic QC to the metadata (nUMI, mitochondrial %)</p></li>
</ul>
<dl class="simple">
<dt>parameters:</dt><dd><ul class="simple">
<li><p>path_input_fullres_image (str) - path of the fullres image</p></li>
<li><p>path_input_data (str) - folder with outs of the Visium. typically square_002um                             (with h5 files and with folders filtered_feature_bc_matrix, spatial)</p></li>
<li><p>path_output (str) - path where to save plots and files</p></li>
<li><p>name (str) - name of the instance</p></li>
<li><p>crop_images (bool) - crop the regions outside of the spots cover area</p></li>
<li><p>properties (dict) - can be any metadata, such as organism, organ, sample_id</p></li>
<li><p>on_tissue_only (bool) - remove spots that are not classified as “on tissue”?</p></li>
<li><p>min_reads_in_spot (int) - filter out spots with less than X UMIs</p></li>
<li><p>min_reads_gene (int) - filter out gene that is present in less than X spots</p></li>
<li><p>fluorescence - either False for H&amp;E, or a dict of channel names and colors. color can be None.</p></li>
<li><p>plot_qc (bool) - plot QC when object is being created</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="HiVis.HiVis">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">HiVis.</span></span><span class="sig-name descname"><span class="pre">HiVis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_fullres</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_highres</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_lowres</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scalefactor_json</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_output</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">properties</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">agg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fluorescence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_qc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/HiVis.html#HiVis"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#HiVis.HiVis" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="HiVis.HiVis.recolor">
<span class="sig-name descname"><span class="pre">recolor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fluorescence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalization_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'percentile'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/HiVis.html#HiVis.recolor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#HiVis.HiVis.recolor" title="Link to this definition"></a></dt>
<dd><p>Recolors a flurescence image
parameters:</p>
<blockquote>
<div><ul class="simple">
<li><p>fluorescence is either list of colors or dict {channel: color…}. color can be None.</p></li>
<li><p>normalization_method - {“percentile”, “histogram”,”clahe”,”sqrt” or None for minmax}</p></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="HiVis.HiVis.qc">
<span class="sig-name descname"><span class="pre">qc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(10,</span> <span class="pre">10)</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/HiVis.html#HiVis.qc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#HiVis.HiVis.qc" title="Link to this definition"></a></dt>
<dd><p>plots basic QC (nUMI, mitochondrial %)</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="HiVis.HiVis.add_mask">
<span class="sig-name descname"><span class="pre">add_mask</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mask_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Paired'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/HiVis.html#HiVis.add_mask"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#HiVis.HiVis.add_mask" title="Link to this definition"></a></dt>
<dd><p>assigns each spot a value based on mask (image).
parameters:</p>
<blockquote>
<div><ul class="simple">
<li><p>mask_path (str) - path to mask image</p></li>
<li><p>name (str) - name of the mask (that will be called in the metadata)</p></li>
<li><p>plot (bool) - plot the mask?</p></li>
<li><p>cmap (str) - colormap for plotting</p></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="HiVis.HiVis.add_annotations">
<span class="sig-name descname"><span class="pre">add_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">measurements</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/HiVis.html#HiVis.add_annotations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#HiVis.HiVis.add_annotations" title="Link to this definition"></a></dt>
<dd><p>Adds annotations made in Qupath (geojson)
parameters:</p>
<blockquote>
<div><ul class="simple">
<li><p>path (str) - path to geojson file</p></li>
<li><p>name (str) - name of the annotation (that will be called in the obs)</p></li>
<li><p>measurements (bool) - include measurements columns?</p></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="HiVis.HiVis.dge">
<span class="sig-name descname"><span class="pre">dge</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">column</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'wilcox'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">two_sided</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umi_thresh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/HiVis.html#HiVis.dge"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#HiVis.HiVis.dge" title="Link to this definition"></a></dt>
<dd><p>Runs differential gene expression analysis between two groups.
parameters:</p>
<blockquote>
<div><ul class="simple">
<li><p>column (str) - which column in obs has the groups classification</p></li>
<li><p>group1 - specific value in the “column”</p></li>
<li><p>group2 - specific value in the “column”.                        if None,will run agains all other values, and will be called “rest”</p></li>
<li><p>method - either “wilcox” or “t_test”</p></li>
<li><p>two_sided (bool) - if one sided, will give the pval for each group,                           and the minimal of both groups (which will also be FDR adjusted)</p></li>
<li><p>umi_thresh (int) - use only spots with more UMIs than this number</p></li>
<li><p>inplace (bool) - modify the adata.var with log2fc, pval and expression columns?</p></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="HiVis.HiVis.add_agg">
<span class="sig-name descname"><span class="pre">add_agg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata_agg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/HiVis.html#HiVis.add_agg"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#HiVis.HiVis.add_agg" title="Link to this definition"></a></dt>
<dd><p>Creates and adds Aggregation to the HiVis instance. Can be accessed by self.agg[name].
For example single-cells, tissue structures.
parameters:</p>
<blockquote>
<div><ul class="simple">
<li><p>adata_agg (ad.AnnData) - anndata containing aggregations</p></li>
<li><p>name (str) - name of the agg</p></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="HiVis.HiVis.agg_stardist">
<span class="sig-name descname"><span class="pre">agg_stardist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'SC'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obs2add</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obs2agg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/HiVis.html#HiVis.agg_stardist"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#HiVis.HiVis.agg_stardist" title="Link to this definition"></a></dt>
<dd><p>Adds Aggregation object to self.agg[name], based on CSV output of Stardist pipeline.
Parameters:</p>
<blockquote>
<div><ul class="simple">
<li><p>input_df (pd.DataFrame) - output of Stardist pipeline</p></li>
<li><p>name (str) - name to store the Aggregation in. Can be acessed via HiVis.agg[name]</p></li>
<li><p>obs2agg - what obs to aggregate from the HiVis.                         Can be a list of column names. numeric columns will be summed, categorical will be the mode.                         Can be a dict specifying the aggregation function.                         examples: {“value_along_axis”:np.median} or {“value_along_axis”:[np.median,np.mean]}</p></li>
<li><p>obs2add (list) - which columns from input_df should be copied to the Aggregation.adata.obs</p></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="HiVis.HiVis.add_meta">
<span class="sig-name descname"><span class="pre">add_meta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type_</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'obs'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/HiVis.html#HiVis.add_meta"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#HiVis.HiVis.add_meta" title="Link to this definition"></a></dt>
<dd><p>Adds a vector to metadata (obs or var)
parameters:</p>
<blockquote>
<div><ul class="simple">
<li><p>name (str) - name of metadata</p></li>
<li><p>values (array like) - values to add</p></li>
<li><p><a href="#id3"><span class="problematic" id="id4">type_</span></a> - either “obs” or “var”</p></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="HiVis.HiVis.update_meta">
<span class="sig-name descname"><span class="pre">update_meta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type_</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'obs'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/HiVis.html#HiVis.update_meta"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#HiVis.HiVis.update_meta" title="Link to this definition"></a></dt>
<dd><p>Updates values in metadata (obs or var)
parameters:</p>
<blockquote>
<div><ul class="simple">
<li><p>name (str) - name of metadata</p></li>
<li><p>values (dict) -{old_value:new_value}</p></li>
<li><p><a href="#id5"><span class="problematic" id="id6">type_</span></a> - either “obs” or “var”</p></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="HiVis.HiVis.pseudobulk">
<span class="sig-name descname"><span class="pre">pseudobulk</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/HiVis.html#HiVis.pseudobulk"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#HiVis.HiVis.pseudobulk" title="Link to this definition"></a></dt>
<dd><p>Returns the gene expression for each group in a single obs.
If “by” is None, will return the mean expression of every gene.
Else, will return a dataframe, each column is a value in “by” (for example cluster), rows are genes.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="HiVis.HiVis.noise_mean_curve">
<span class="sig-name descname"><span class="pre">noise_mean_curve</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">plot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">signif_thresh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.95</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/HiVis.html#HiVis.noise_mean_curve"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#HiVis.HiVis.noise_mean_curve" title="Link to this definition"></a></dt>
<dd><p>Generates a noise-mean curve of the data.
Parameters:</p>
<blockquote>
<div><ul class="simple">
<li><p>plot (bool) - plot the curve?</p></li>
<li><p>layer - which layer in the anndata to use</p></li>
<li><p>signif_thresh (float) - for plotting, add text for genes in this residual precentile</p></li>
<li><p>inplace (bool) - add the mean_expression, cv and residuals to VAR?</p></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="HiVis.HiVis.cor">
<span class="sig-name descname"><span class="pre">cor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">what</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">self_corr_value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normilize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/HiVis.html#HiVis.cor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#HiVis.HiVis.cor" title="Link to this definition"></a></dt>
<dd><p>Calculates gene(s) correlation.
Parameters:</p>
<blockquote>
<div><ul class="simple">
<li><p>what (str or list) - if str, computes Spearman correlation of a given gene with all genes.                                     if list, will compute correlation between all genes in the list</p></li>
<li><p>self_corr_value - replace the correlation of the gene with itself by this value</p></li>
<li><p>normilize (bool) - normilize expression before computing correlation?</p></li>
<li><p>layer (str) - which layer in the anndata to use</p></li>
<li><p>inplace (bool) - add the correlation to VAR?</p></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="HiVis.HiVis.export_h5">
<span class="sig-name descname"><span class="pre">export_h5</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/HiVis.html#HiVis.export_h5"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#HiVis.HiVis.export_h5" title="Link to this definition"></a></dt>
<dd><p>Exports the adata as h5ad.
Parameters:</p>
<blockquote>
<div><ul class="simple">
<li><p>path (str) - path to save the h5 file. If None, will save to path_output</p></li>
<li><p>force (bool) - save file even if it allready exists</p></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="HiVis.HiVis.export_images">
<span class="sig-name descname"><span class="pre">export_images</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/HiVis.html#HiVis.export_images"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#HiVis.HiVis.export_images" title="Link to this definition"></a></dt>
<dd><p>Exports full,high and low resolution images
Parameters:</p>
<blockquote>
<div><ul class="simple">
<li><p>path (str) - path to save the image files. If None, will save to path_output</p></li>
<li><p>force (bool)- save files even if they allready exists</p></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="HiVis.HiVis.get">
<span class="sig-name descname"><span class="pre">get</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">what</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cropped</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/HiVis.html#HiVis.get"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#HiVis.HiVis.get" title="Link to this definition"></a></dt>
<dd><p>get a vector from data (a gene) or metadata (from obs or var). or subset the object.
parameters:</p>
<blockquote>
<div><ul class="simple">
<li><p>what - if string, will get data or metadata. else, will return a new HiVis object that is spliced.                      the splicing is passed to the self.adata</p></li>
<li><p>cropped - get the data from the adata_cropped after crop() or plotting methods?</p></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="HiVis.HiVis.subset">
<span class="sig-name descname"><span class="pre">subset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">what</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(slice(None,</span> <span class="pre">None,</span> <span class="pre">None),</span> <span class="pre">slice(None,</span> <span class="pre">None,</span> <span class="pre">None))</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_empty_pixels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">crop_agg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/HiVis.html#HiVis.subset"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#HiVis.HiVis.subset" title="Link to this definition"></a></dt>
<dd><p>Create a new HiVis objects based on adata subsetting.
parameters:</p>
<blockquote>
<div><ul class="simple">
<li><dl class="simple">
<dt>what (tuple) - tuple of two elements. slicing instruction for adata. examples:</dt><dd><ul>
<li><p>(slice(None), slice(None)): Select all spots and all genes.</p></li>
<li><p>([0, 1, 2], slice(None)): Select the first three spots and all genes.</p></li>
<li><p>(slice(None), [‘GeneA’, ‘GeneB’]): Select all spots and specific genes.</p></li>
<li><p>(adata.obs[‘obs1’] == ‘value’, slice(None)): Select spots where 
the ‘obs1’ column in adata.obs is ‘value’, and all genes.</p></li>
</ul>
</dd>
</dl>
</li>
<li><p>remove_empty_pixels (bool) - if True, the images will only contain pixels under visium spots</p></li>
<li><p>crop_agg (bool) - crop agg objects? If False, plotting of aggregations might break.</p></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="HiVis.HiVis.remove_pixels">
<span class="sig-name descname"><span class="pre">remove_pixels</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">column</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">marging</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/HiVis.html#HiVis.remove_pixels"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#HiVis.HiVis.remove_pixels" title="Link to this definition"></a></dt>
<dd><p>Removes pixels in images, based on adata.obs[column].isin(values).
parameters:</p>
<blockquote>
<div><ul class="simple">
<li><p>marging (int) - how many pixels to extend the removed pixels.</p></li>
</ul>
</div></blockquote>
<p>returns new HiVis object.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="HiVis.HiVis.head">
<span class="sig-name descname"><span class="pre">head</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/HiVis.html#HiVis.head"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#HiVis.HiVis.head" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="HiVis.HiVis.shape">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">shape</span></span><a class="headerlink" href="#HiVis.HiVis.shape" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="HiVis.HiVis.columns">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">columns</span></span><a class="headerlink" href="#HiVis.HiVis.columns" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="HiVis.HiVis.rename">
<span class="sig-name descname"><span class="pre">rename</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">new_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_out_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">full</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/HiVis.html#HiVis.rename"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#HiVis.HiVis.rename" title="Link to this definition"></a></dt>
<dd><p>Renames the object and changes the path_output.
If full is False, the name will be added to the current (previous) name</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="HiVis.HiVis.update">
<span class="sig-name descname"><span class="pre">update</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">agg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/HiVis.html#HiVis.update"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#HiVis.HiVis.update" title="Link to this definition"></a></dt>
<dd><p>Updates the methods in the instance. Should be used after modifying the source code in the class</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="HiVis.HiVis.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/HiVis.html#HiVis.copy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#HiVis.HiVis.copy" title="Link to this definition"></a></dt>
<dd><p>Creates a deep copy of the instance</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="HiVis.HiVis.save">
<span class="sig-name descname"><span class="pre">save</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/HiVis.html#HiVis.save"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#HiVis.HiVis.save" title="Link to this definition"></a></dt>
<dd><p>Saves the instance in pickle format.
If no path specified, will save in the path_output as the name of the instance</p>
</dd></dl>

</dd></dl>

</section>
<section id="module-Aggregation">
<span id="aggregation"></span><h2>Aggregation<a class="headerlink" href="#module-Aggregation" title="Link to this heading"></a></h2>
<p>Aggregation of spots from HiVis</p>
<dl class="py class">
<dt class="sig sig-object py" id="Aggregation.Aggregation">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Aggregation.</span></span><span class="sig-name descname"><span class="pre">Aggregation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">hiviz_instance</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata_agg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">geojson_agg_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Aggregation.html#Aggregation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#Aggregation.Aggregation" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="Aggregation.Aggregation.import_geometry">
<span class="sig-name descname"><span class="pre">import_geometry</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">geojson_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">object_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cell'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Aggregation.html#Aggregation.import_geometry"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#Aggregation.Aggregation.import_geometry" title="Link to this definition"></a></dt>
<dd><p>Adds “geometry” column to self.adata.obs, based on Geojson exported from Qupath.
parameters:</p>
<blockquote>
<div><ul class="simple">
<li><p>geojson_path (str) - path to geojson file</p></li>
<li><p>object_type (str) - which “objectType” to merge from the geojson</p></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Aggregation.Aggregation.merge">
<span class="sig-name descname"><span class="pre">merge</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pca</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hvg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Aggregation.html#Aggregation.merge"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#Aggregation.Aggregation.merge" title="Link to this definition"></a></dt>
<dd><p>Merge info from an anndata to self.adata, in case genes have been filtered.
parameters:</p>
<blockquote>
<div><ul class="simple">
<li><p>adata (as.AnnData) - anndata where to get the values from</p></li>
<li><p>obs - single string or list of obs to merge</p></li>
<li><p>var - single string or list of var to merge</p></li>
<li><p>umap (bool) - add umap to OBSM, and UMAP coordinates to obs?</p></li>
<li><p>pca (bool) - add PCA to OBSM?</p></li>
<li><p>hvg (bool) - add highly variable genes to vars?</p></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Aggregation.Aggregation.get">
<span class="sig-name descname"><span class="pre">get</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">what</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cropped</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">geometry</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Aggregation.html#Aggregation.get"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#Aggregation.Aggregation.get" title="Link to this definition"></a></dt>
<dd><p>get a vector from data (a gene) or metadata (from obs or var). or subset the object.
parameters:</p>
<blockquote>
<div><ul class="simple">
<li><p>what - if string, will get data or metadata.                      else, will return a new Aggregation object that is spliced.                      the splicing is passed to the self.adata.</p></li>
<li><p>cropped (bool) - get the data from the adata_cropped after crop() or plotting methods?</p></li>
<li><p>geometry (bool) - include only objects which have geometry</p></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Aggregation.Aggregation.subset">
<span class="sig-name descname"><span class="pre">subset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">what</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(slice(None,</span> <span class="pre">None,</span> <span class="pre">None),</span> <span class="pre">slice(None,</span> <span class="pre">None,</span> <span class="pre">None))</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Aggregation.html#Aggregation.subset"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#Aggregation.Aggregation.subset" title="Link to this definition"></a></dt>
<dd><p>Create a new Aggregation object based on adata subsetting.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Aggregation.Aggregation.pseudobulk">
<span class="sig-name descname"><span class="pre">pseudobulk</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Aggregation.html#Aggregation.pseudobulk"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#Aggregation.Aggregation.pseudobulk" title="Link to this definition"></a></dt>
<dd><p>Returns the gene expression for each group in a single obs.
If “by” is None, will return the mean expression of every gene.
Else, will return a dataframe, each column is a value in “by” (for example cluster), rows are genes.
layer - which layer in adata to use.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Aggregation.Aggregation.smooth">
<span class="sig-name descname"><span class="pre">smooth</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">what</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">radius</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'median'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_col_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Aggregation.html#Aggregation.smooth"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#Aggregation.Aggregation.smooth" title="Link to this definition"></a></dt>
<dd><p>Applies median smoothing to the specified column in adata.obs using spatial neighbors.
parameters:</p>
<blockquote>
<div><ul class="simple">
<li><p>what (str) - what to smooth. either a gene name or column name from self.adata.obs</p></li>
<li><p>radius (float) - in microns</p></li>
<li><p>method - [“mode”,”median”, “mean”, “gaussian”, “log”]</p></li>
<li><p>new_col_name (str) - Optional custom name for the output column.</p></li>
</ul>
<p><a href="#id1"><span class="problematic" id="id2">**</span></a>kwargs - Additional parameters for specific methods (e.g., sigma for gaussian, offset for log).</p>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Aggregation.Aggregation.noise_mean_curve">
<span class="sig-name descname"><span class="pre">noise_mean_curve</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">plot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">signif_thresh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.95</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Aggregation.html#Aggregation.noise_mean_curve"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#Aggregation.Aggregation.noise_mean_curve" title="Link to this definition"></a></dt>
<dd><p>Generates a noise-mean curve of the data.
Parameters:</p>
<blockquote>
<div><ul class="simple">
<li><p>plot (bool) - plot the curve?</p></li>
<li><p>layer (str) - which layer in the anndata to use</p></li>
<li><p>signif_thresh (float) - for plotting, add text for genes in this residual precentile</p></li>
<li><p>inplace (bool) - add the mean_expression, cv and residuals to VAR?</p></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Aggregation.Aggregation.cor">
<span class="sig-name descname"><span class="pre">cor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">what</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">self_corr_value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normilize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Aggregation.html#Aggregation.cor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#Aggregation.Aggregation.cor" title="Link to this definition"></a></dt>
<dd><p>Calculates gene(s) correlation.
Parameters:</p>
<blockquote>
<div><ul class="simple">
<li><p>what (str or list) - if str, computes Spearman correlation of a given gene with all genes.                                     if list, will compute correlation between all genes in the list</p></li>
<li><p>self_corr_value - replace the correlation of the gene with itself by this value</p></li>
<li><p>normilize (bool) - normilize expression before computing correlation?</p></li>
<li><p>layer (str) - which layer in the anndata to use</p></li>
<li><p>inplace (bool) - add the correlation to VAR?</p></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Aggregation.Aggregation.sync">
<span class="sig-name descname"><span class="pre">sync</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">what</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Aggregation.html#Aggregation.sync"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#Aggregation.Aggregation.sync" title="Link to this definition"></a></dt>
<dd><p>Transfers metadata assignment from the Aggregation to the spots.
what - obs column name to pass to HiViz object</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Aggregation.Aggregation.export_h5">
<span class="sig-name descname"><span class="pre">export_h5</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Aggregation.html#Aggregation.export_h5"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#Aggregation.Aggregation.export_h5" title="Link to this definition"></a></dt>
<dd><p>Exports the adata. 
* path (str) - path to save the h5 file. If None, will save to path_output</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Aggregation.Aggregation.dge">
<span class="sig-name descname"><span class="pre">dge</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">column</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'wilcox'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">two_sided</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umi_thresh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Aggregation.html#Aggregation.dge"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#Aggregation.Aggregation.dge" title="Link to this definition"></a></dt>
<dd><p>Runs differential gene expression analysis between two groups.
Values will be saved in self.var: expression_mean, log2fc, pval
parameters:</p>
<blockquote>
<div><ul class="simple">
<li><p>column - which column in obs has the groups classification</p></li>
<li><p>group1 - specific value in the “column”</p></li>
<li><p>group2 - specific value in the “column”.                        if None,will run agains all other values, and will be called “rest”</p></li>
<li><p>method - either “wilcox” or “t_test”</p></li>
<li><p>two_sided - if one sided, will give the pval for each group,                           and the minimal of both groups (which will also be FDR adjusted)</p></li>
<li><p>umi_thresh - use only spots with more UMIs than this number</p></li>
<li><p>expression - function F {mean, mean, max} F(mean(group1),mean(group2))</p></li>
<li><p>inplace - modify the adata.var with log2fc, pval and expression columns?</p></li>
<li><p>layer (str) - which layer in the anndata to use</p></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Aggregation.Aggregation.shape">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">shape</span></span><a class="headerlink" href="#Aggregation.Aggregation.shape" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Aggregation.Aggregation.combine">
<span class="sig-name descname"><span class="pre">combine</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Aggregation.html#Aggregation.combine"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#Aggregation.Aggregation.combine" title="Link to this definition"></a></dt>
<dd><p>Combines two Aggregation objects into a single adata</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Aggregation.Aggregation.update">
<span class="sig-name descname"><span class="pre">update</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/Aggregation.html#Aggregation.update"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#Aggregation.Aggregation.update" title="Link to this definition"></a></dt>
<dd><p>Updates the methods in the instance. Should be used after modifying the source code in the class</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Aggregation.Aggregation.head">
<span class="sig-name descname"><span class="pre">head</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Aggregation.html#Aggregation.head"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#Aggregation.Aggregation.head" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Aggregation.Aggregation.columns">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">columns</span></span><a class="headerlink" href="#Aggregation.Aggregation.columns" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Aggregation.Aggregation.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/Aggregation.html#Aggregation.copy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#Aggregation.Aggregation.copy" title="Link to this definition"></a></dt>
<dd><p>Creates a deep copy of the instance</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Aggregation.Aggregation.rename">
<span class="sig-name descname"><span class="pre">rename</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">new_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_out_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">full</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Aggregation.html#Aggregation.rename"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#Aggregation.Aggregation.rename" title="Link to this definition"></a></dt>
<dd><p>Renames the object and changes the path_output.
If full is False, the name will be added to the current (previous) name</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Aggregation.Aggregation.export_to_matlab">
<span class="sig-name descname"><span class="pre">export_to_matlab</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Aggregation.html#Aggregation.export_to_matlab"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#Aggregation.Aggregation.export_to_matlab" title="Link to this definition"></a></dt>
<dd><p>exports gene names, data (sparse matrix) and metadata to a .mat file</p>
</dd></dl>

</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="HD Integrated Visium Interactive Suite (HiVis)" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="test.html" class="btn btn-neutral float-right" title="Aggregation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Roy Novoselsky.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>